pkg/kubelet/cm/internal_container_lifecycle_windows.go

```diff
@@ -20,30 +20,160 @@ limitations under the License.
+			affinity, err := winstats.GetCPUsforNUMANode(uint16(numaNode))
+			if err != nil {
+				return fmt.Errorf("failed to get CPUs for NUMA node %d: %v", numaNode, err)
 			}
+			allNumaNodeCPUs = append(allNumaNodeCPUs, *affinity)
+		}
+
+		// Create sets from integer representation of CPUs ((group * 64) + processorId for each bit set in the affinity mask)
+		cpuManagerAffinityCPUSet := computeCPUSet(convertToGroupAffinities(allocatedCPUs))
+		numaNodeAffinityCPUSet := computeCPUSet(allNumaNodeCPUs)
+
+		// Determine which set of CPUs to use using the following logic outlined in the KEP:
+		// Case 1: CPU manager selects more CPUs than those availble in the NUMA nodes selected by the memory manager
+		// Case 2: CPU manager selects fewer CPUs, and they all fall within the CPUs available in the NUMA nodes selected by the memory manager
+		// Case 3: CPU manager selects fewer CPUs, but some are outside of the CPUs available in the NUMA nodes selected by the memory manager
 
-			containerConfig.Windows.Resources.AffinityCpus = cpuGroupAffinities
+		if len(cpuManagerAffinityCPUSet) > len(numaNodeAffinityCPUSet) {
+			// Case 1, use CPU manager selected CPUs
+			finalCPUSet = cpuManagerAffinityCPUSet
+		} else if subset(cpuManagerAffinityCPUSet, numaNodeAffinityCPUSet) {
+			// case 2, use CPU manager selected CPUs
+			finalCPUSet = cpuManagerAffinityCPUSet
+		} else {
+			// Case 3, merge CPU manager and memory manager selected CPUs
+			finalCPUSet = mergeSets(cpuManagerAffinityCPUSet, numaNodeAffinityCPUSet)
+		}
+	} else if i.cpuManager != nil && !allocatedCPUs.IsEmpty() {
+		// Only CPU manager is enabled, use CPU manager selected CPUs
+		finalCPUSet = computeCPUSet(convertToGroupAffinities(allocatedCPUs))
+	} else if i.memoryManager != nil && !allocatedCPUs.IsEmpty() {
+		// Only memory manager is enabled, use CPUs associated with selected NUMA nodes
+		var allNumaNodeCPUs []winstats.GroupAffinity
+		for _, numaNode := range sets.List(numaNodes) {
+			affinity, err := winstats.GetCPUsforNUMANode(uint16(numaNode))
+			if err != nil {
+				return fmt.Errorf("failed to get CPUs for NUMA node %d: %v", numaNode, err)
+			}
+			allNumaNodeCPUs = append(allNumaNodeCPUs, *affinity)
 		}
+		finalCPUSet = computeCPUSet(allNumaNodeCPUs)
 	}
+
+	// Set CPU group affinities in the container config
+	if finalCPUSet != nil {
+		var cpusToGroupAffinities []*runtimeapi.WindowsCpuGroupAffinity
+		for group, mask := range groupMasks(finalCPUSet) {
+			cpusToGroupAffinities = append(cpusToGroupAffinities, &runtimeapi.WindowsCpuGroupAffinity{
+				CpuGroup: uint32(group),
+				CpuMask:  uint64(mask),
+			})
+		}
+		containerConfig.Windows.Resources.AffinityCpus = cpusToGroupAffinities
+	}
+
+	// return nil if no CPUs were selected
 	return nil
 }
+
+// computeCPUSet returns a map of CPU IDs to an empty struct based on the provided group affinities
+func computeCPUSet(affinities []winstats.GroupAffinity) map[int]struct{} {
```

Comment:
can we use sets from `apimachinery/util` (or its most current replacement like perhaps `k8s.io/utils`)?

Reply:
yes, i'll make some updates

Reply:
Switched to apimachinery/util since it had everything I needed.